### PR TITLE
add Socksify so we can use SAFE_PROXY with socks+ssh

### PR DESCRIPTION
**Pull request by ramonskie**
We want to be able to use socks+ssh on vault
if it is behind a proxy and you have a jumpbox
just like credhub and bosh

https://github.com/starkandwayne/safe/issues/132

### CODE REVIEW

**Comment by thomasmmitchell** [comment]
```
~/go_workspace/src/github.com/starkandwayne/safe → go build
# github.com/starkandwayne/safe/vendor/github.com/cloudfoundry/socks5-proxy
vendor/github.com/cloudfoundry/socks5-proxy/socks5_proxy.go:82:47: undefined: ssh.FixedHostKey
vendor/github.com/cloudfoundry/socks5-proxy/ssh_client_config.go:8:54: undefined: ssh.HostKeyCallback
vendor/github.com/cloudfoundry/socks5-proxy/ssh_client_config.go:10:10: unknown field 'Timeout' in struct literal of type ssh.ClientConfig
```

Doesn't quite build on my machine yet.

**Comment by thomasmmitchell** [comment]
This does have the unintuitive quality that if the SAFE_PROXY string is malformed, it just ignores it and acts as if no proxy was specified.

**Comment by thomasmmitchell** [comment]
We also need a way to document the usage of this command. Probably need a command that lists the available environment variables and their usage.

**Comment by ramonskie** [comment]
but we haven't done that with `http_proxy` var as well right?

**Comment by ramonskie** [comment]
i meant that i could not find anything in the documentation related about http_proxy.
but in any case. where do you want this documented?
in the readme or in safe itself

**Comment by ramonskie** [comment]
@thomasmmitchell can we merge this?

**Comment by thomasmmitchell** [comment]
Yeah, it's been on my radar as something to test and integrate once I got done with the versioning stuff (didn't want _more_ stuff to test at the time). That's pretty much just wrapped. I've got to do something else so that Dennis can be unblocked on genesis, and then I'll try to use some time to test this so it can get merged.

**Comment by thomasmitchell** [comment]
So, I was taking a dive into this yesterday. I want to integrate this into the targets system - basically, define a target with a proxy. I also sort of want to split the all magical SAFE_PROXY variable into HTTPS_PROXY, SSH_PROXY, and SSH_PROXY_PRIVATE_KEY variables (which would serve as overrides for the selected target). I feel there's a little too much magic in the all-in-one variable that makes its less intuitive for people who didn't write it.

The core code looks good - just a bit of shuffling. It's something I can handle when I get the time.

**Comment by ramonskie** [comment]
it seems a bit of standard practice to use the all in one variable
as seen with bosh, credhub, fly, uaa etc

**Comment by jhunt** [comment]
Those examples are all Pivotal tooling; I would expect them to be internally consistent.  Is this a practice that other tools use?  I know most things dealing with HTTP and proxying separate out `http_proxy` from `https_proxy` and don't fallback to one or the other.

File: vault/proxy.go [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	socks5 "github.com/armon/go-socks5"
+	isatty "github.com/mattn/go-isatty"
```
Comment (jhunt) [inline vault/proxy.go]: you shouldn't need the aliases here; IIRC, golang automagically strips the `go-` prefix from any import terminal names.

File: vault/proxy.go [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+	allProxy := getEnvironmentVariable("SAFE_ALL_PROXY", "safe_all_proxy")
+	if allProxy != "" {
+		httpProxy = allProxy
+		httpsProxy = allProxy
+	}
+
+	noProxy := getEnvironmentVariable("NO_PROXY", "no_proxy")
+
+	knownHostsFile := getEnvironmentVariable("KNOWN_HOSTS_FILE", "known_hosts_file")
+	skipHostKeyString := getEnvironmentVariable("SKIP_HOST_KEY_VALIDATION", "skip_host_key_validation")
```
Comment (jhunt) [inline vault/proxy.go]: Are these conventional env vars, or are they specific to safe?  If the latter, they should probably be prefixed with `SAFE_` / `safe_`.

File: vault/proxy.go [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+	skipHostKeyValidation := skipHostKeyString != "" && skipHostKeyString != "false"
```
Comment (jhunt) [inline vault/proxy.go]: We should also treat "0" and "no" as falsey.

File: vault/proxy.go [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+		tunnelsToOpen = append(tunnelsToOpen, &httpProxy)
+	}
+
+	if strings.HasPrefix(httpsProxy, "ssh+socks5://") {
+		tunnelsToOpen = append(tunnelsToOpen, &httpsProxy)
+	}
+
+	for i := range tunnelsToOpen {
+		//If we haven't already opened a proxy for this...
+		if i == 0 || *(tunnelsToOpen[i]) != *(tunnelsToOpen[i-1]) {
```
Comment (jhunt) [inline vault/proxy.go]: Can this be more cleanly (and clearly) done with a map?

File: vault/proxy.go [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+		},
+	}, nil
+}
+
+func getEnvironmentVariable(variables ...string) string {
+	var ret string
+	for _, v := range variables {
+		ret := os.Getenv(v)
+		if ret != "" {
+			break
```
Comment (jhunt) [inline vault/proxy.go]: Can't you just `return ret` here, and then `return ""` at the end of the function?  Also, FWIW, os.LookupEnv returns `(string, bool)` so you can determine if the value is set; not sure if we have covered all the edge cases where users explicitly set the env var to "", and expected different behavior than when they didn't set it in the first place.

File: vault/proxy.go [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+//StartSSHTunnel makes an SSH connection according to the given config. It
+// returns an SSH client if it was successful and an error otherwise.
+func StartSSHTunnel(conf SOCKS5SSHConfig) (*ssh.Client, error) {
+	hostKeyCallback := ssh.InsecureIgnoreHostKey()
+	var err error
+
+	if !conf.SkipHostKeyValidation {
+		if conf.KnownHostsFile == "" {
+			conf.KnownHostsFile = fmt.Sprintf("%s/.ssh/known_hosts", os.Getenv("HOME"))
```
Comment (jhunt) [inline vault/proxy.go]: What do we do if `$HOME` isn't set?

File: vault/proxy.go [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+	if err != nil {
+		return "", fmt.Errorf("Error starting local SOCKS5 server: %s", err)
+	}
+
+	socks5Listener, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		return "", fmt.Errorf("Error starting local SOCKS5 server: %s", err)
+	}
+
+	//TODO: Put this on another thread. Get error somehow?
```
Comment (jhunt) [inline vault/proxy.go]: Should we resolve this TODO?

File: vault/proxy.go (hunk 1 above)
Comment (thomasmitchell) [reply vault/proxy.go]: goimports does this automatically. Gonna leave this to avoid jitter in the future.

Comment (jhunt) [reply vault/proxy.go]: dumb, but okay 😄

File: vault/proxy.go (hunk 2 above)
Comment (jhunt) [reply vault/proxy.go]: 👍

File: vault/proxy.go (hunk 3 above)
Comment (jhunt) [reply vault/proxy.go]: :+1:

File: vault/proxy.go (hunk 4 above)
Comment (jhunt) [reply vault/proxy.go]: :+1:

File: vault/proxy.go (hunk 5 above)
Comment (jhunt) [reply vault/proxy.go]: :+1:

File: vault/proxy.go (hunk 6 above)
Comment (jhunt) [reply vault/proxy.go]: :+1:

File: vault/proxy.go (hunk 7 above)
Comment (jhunt) [reply vault/proxy.go]: :+1:

**Pull request merged**